### PR TITLE
NIFI-3518 Create a Morphlines processor

### PR DESCRIPTION
**Pull request by WilliamNouet**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

For all changes:

[Y] Is there a JIRA ticket associated with this PR? Is it referenced
in the commit message?

[Y] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

[Y] Has your PR been rebased against the latest commit within the target branch (typically master)?

[Y] Is your initial contribution a single, squashed commit?

For code changes:

[Y] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
[Y] Have you written or updated unit tests to verify your changes?
[N/A] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under ASF 2.0?
[N/A] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
[N/A] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
[N/A] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?
For documentation related changes:

[N/A] Have you ensured that format looks appropriate for the output in which it is rendered?
Note:

Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by WilliamNouet** [comment]
This follows the work done on https://github.com/apache/nifi/pull/1529. Opened a new PR for cleaner work.

**Comment by WilliamNouet** [comment]
@JPercivall @trixpan Any update on this PR?

**Comment by JPercivall** [comment]
Hey @WilliamNouet, sorry for not giving an update. Been too busy with other obligations. I'll try to get back to this when I have more free time.

**Comment by JPercivall** [comment]
Sorry for the delay @WilliamNouet . I am going to look at it now.

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/pom.xml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.11</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.kitesdk</groupId>
+            <artifactId>kite-morphlines-core</artifactId>
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/pom.xml]: All of these dependencies are missing versions. I can't build it at all.

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-nar/pom.xml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.apache.com</groupId>
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-nar/pom.xml]: This is wrong, should be "org.apache.nifi".

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ImplementMorphlines.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+            .build();
+
+
+    private File morphLinesFile;
+    private String morphLinesId;
+    private String morphlinesOutputField;
+    private PropertyValue morphLinesFileProperty;
+    private PropertyValue morphLinesIdProperty;
+    private PropertyValue morphLinesOutputField;
+    private MorphlineContext morphlineContext;
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ImplementMorphlines.java]: You didn't address my comment[1] on these variables. Please do so.

[1] https://github.com/apache/nifi/pull/1529#discussion_r102964711

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-nar/pom.xml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        <artifactId>nifi-morphlines-bundle</artifactId>
+        <version>1.2.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>nifi-morphlines-nar</artifactId>
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-nar/pom.xml]: In order for the nifi-assembly pom to know to include it into the final package, you need to add it as a dependency of nifi-assembly/pom.xml.

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-nar/pom.xml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    <packaging>nar</packaging>
+    <properties>
+        <maven.javadoc.skip>true</maven.javadoc.skip>
+        <source.skip>true</source.skip>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-morphlines-processors</artifactId>
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-nar/pom.xml]: You need to add a license/notice file for this nar and add any needed additions to the nifi-assembly license/notice files. Every dependency you bring in (including their transitive deps) need to be accounted for.

You can see everything you're bringing in by running "mvn dependency:tree -Dverbose".

Our licensing guide is here: https://nifi.apache.org/licensing-guide.html

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ImplementMorphlines.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.*;
+import java.util.stream.*;
+import java.util.concurrent.atomic.*;
+import org.apache.nifi.annotation.lifecycle.OnStopped;
+import org.apache.nifi.processor.exception.*;
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ImplementMorphlines.java]: Please run "mvn clean install -Pcontrib-check" in order to see all your checkstyle issues.

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ImplementMorphlines.java [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+@Tags({"kitesdk", "morphlines", "ETL", "HDFS", "avro", "Solr", "HBase"})
+@CapabilityDescription("Implements Morphlines (http://kitesdk.org/docs/1.1.0/morphlines/) framework, which performs in-memory container of transformation commands in oder to perform tasks such as loading, parsing, transforming, or otherwise processing a single record.")
+public class MorphlinesProcessor extends AbstractProcessor {
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ImplementMorphlines.java]: Please change the file name or classname in order to match.

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ImplementMorphlines.java [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+        try{
+            flowFile = session.write(flowFile, new StreamCallback() {
+                @Override
+                public void process(InputStream in, OutputStream out) throws IOException {
+                    StreamUtils.fillBuffer(in, value);
+                    Record record = new Record();
+                    record.put(Fields.ATTACHMENT_BODY, value);
+                    Collector collectorRecord = new Collector();
+                    morphline = new org.kitesdk.morphline.base.Compiler().compile(morphLinesFile, morphLinesId, morphlineContext, collectorRecord);
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/src/main/java/org/apache/nifi/processors/morphlines/ImplementMorphlines.java]: You said before that you couldn't move this out of the onTrigger due to needing the collectorRecord to create and compile it.

According to the Cloudera intro documentation "Example Driver Program". It creates the compiler then processes each record. Can't we do the same here? Relevant section below:

```
  MorphlineContext context = new MorphlineContext.Builder().build();
  Command morphline = new Compiler().compile(configFile, null, context, null);

  // process each input data file
  Notifications.notifyBeginTransaction(morphline);
  for (int i = 1; i < args.length; i++) {
    InputStream in = new FileInputStream(new File(args[i]));
    Record record = new Record();
    record.put(Fields.ATTACHMENT_BODY, in);
    morphline.process(record);
    in.close();
  }
```



[1] http://blog.cloudera.com/blog/2013/07/morphlines-the-easy-way-to-build-and-integrate-etl-apps-for-apache-hadoop/

File: nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/pom.xml [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            <groupId>org.kitesdk</groupId>
+            <artifactId>kite-morphlines-maxmind</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.kitesdk</groupId>
+            <artifactId>kite-morphlines-metrics-servlets</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.kitesdk</groupId>
+            <artifactId>kite-morphlines-useragent</artifactId>
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-morphlines-bundle/nifi-morphlines-processors/pom.xml]: I'm a bit worried that importing all of these dependencies is a too broad in scope. From my understanding not all of them are inherently needed but instead represent a toolbox of things that could potentially be used. 

I'd much prefer we limit this to a core group of functionality that we can't already do in NiFi instead.

**Comment by WilliamNouet** [comment]
Closing this one and opening PR #2028

**Pull request closed without merging**